Add a product version string invalidator

joshuatf opened the pull request:
### Changes proposed in this Pull Request:

This pull request introduces a `ProductVersionStringInvalidator` class that has two purposes:

1. Introduce an `invalidate` method that invalidates the corresponding product entry in the [version strings generator](https://github.com/woocommerce/woocommerce/pull/61600).
2. Capture all the events (hooks) that cause a modification in product information from WooCommerce core and have them execute `invalidate`.

Capturing the product modification events is a prerequisite for introducing backend caching capabilities to the product related REST API endpoints using the `RestApiCache` trait. The corresponding hooks will be handled **only** if the REST API caching feature is enabled **and** the backend caching setting is set (see https://github.com/woocommerce/woocommerce/pull/61986).

We are capturing all the possible product related events and this implies that there will be redundant invalidations for the same product, for example a product creation will trigger an invalidation for the `save_post_product` and `woocommerce_new_product` hooks. This implies that multiple consecutive version string invalidations will happen for the same product, this is harmless but we may want to refine this in a future interation of the invalidator.

Closes (part of) #61151 .

### How to test the changes in this Pull Request

These testing instructions are console-based and quite long, since they cover all the events that `ProductCacheInvalidator` captures as cache invalidating-worthy.

Initial configuration:

- Install a persistent WordPress object cache (recommended: [W3 Total Cache](https://wordpress.org/plugins/w3-total-cache/), you need to go to configuration and explicitly enable the object cache)
- Enable the "REST API caching" feature (_WooCommerce - Settings - Advanced - Features_)
- Enable backend caching for REST API responses (_WooCommerce - Settings - Advanced - REST API caching_)

Now, temporarily modify the `ProductVersionStringInvalidator::invalidate` method to include the following line:

```
do_action( 'temp_product_version_string_invalidated', $product_id );
```

This will help us to verify that actual invalidations are happening. Also add the following auxiliary code:

```php
add_action( 'temp_product_version_string_invalidated', function( $product_id ) {
	echo "--- Product version string invalidated: ID={$product_id}\n";
}, 10, 1);

function create_test_simple_product() {
	$product = new WC_Product_Simple();
	$product->set_name( 'Test Simple Product - ' . time() );
	$product->set_manage_stock( true );
	$product->set_stock_quantity( 100 );
	$product->save();
	return $product;
}

function create_test_variable_product() {
	$parent = new WC_Product_Variable();
	$parent->set_name( 'Test Variable Product - ' . time() );
	$parent->save();

	$variation1 = new WC_Product_Variation();
	$variation1->set_parent_id( $parent->get_id() );
	$variation1->set_regular_price( '10.00' );
	$variation1->save();
	
	$variation2 = new WC_Product_Variation();
	$variation2->set_parent_id( $parent->get_id() );
	$variation2->set_regular_price( '10.00' );
	$variation2->save();

	$parent->set_children( [ $variation1->get_id(), $variation2->get_id() ] );
	$parent->save();

	return [$parent, $variation1, $variation2];
}
```

Now fire `wp shell` and let's roll. Each part is a set of PHP commands to run, with the relevant ones having a comment explaining the expected output.

"For parent and variation" means that the action is fired twice, once for the variation id and once for the parent id; the later has a `variation_id` key in the context array.

#### 1. Manual invalidation

```php
$invalidator = wc_get_container()->get( Automattic\WooCommerce\Internal\Caches\ProductVersionStringInvalidator::class );
$vsg = wc_get_container()->get( Automattic\WooCommerce\Internal\Caches\VersionStringGenerator::class );

// Prints a random version string
$vsg->get_version('product_123');

// Prints: Product version string invalidated: ID=123
$invalidator->invalidate( 123 );

// Shows a different version this time
$vsg->get_version('product_123');
```

#### 2. Product creation and update

```php
// Fires two actions, for hooks 'save_post_product' and 'woocommerce_new_product'
$product = create_test_simple_product();

$product->set_name( 'Test Product - ' . time() );

// Fires two actions, for hooks 'save_post_product' and 'woocommerce_update_product'
$product->save();
```

#### 3. Product trashing and deleting using the posts API

```php
$product_id = $product->get_id();

// Fires two actions, for hooks 'save_post_product'  and 'trashed_post'
wp_trash_post($product_id);

// Fires actions for hooks 'save_post_product', 'woocommerce_update_product' and 'untrashed_post'
wp_untrash_post($product_id);

// Fires one action for the 'delete_post' hook
wp_delete_post($product_id, true);
```

#### 4. Product trashing and deleting using the WooCommerce API

```php
$product = create_test_simple_product();
$product_id = $product->get_id();

// Fires actions for hooks 'save_post_product', 'trashed_post' and 'woocommerce_trash_product'
$product->delete(false);

wp_untrash_post($product_id);

$product = wc_get_product($product_id);

// Fires two actions, for hooks 'woocommerce_before_delete_product' and 'delete_post'
$product->delete(true);
```

#### 5. Operations with variations

```php
// Fires action for the 'woocommerce_new_product_variation' hook twice per variation
// (for parent and variation)
[$parent, $variation1, $variation2] = create_test_variable_product();

$variation1->set_regular_price( '1' );

// Fires action for the 'woocommerce_update_product_variation' hook twice
// (for parent and variation)
$variation1->save();

$variation1_id = $variation1->get_id();

// Fires action for the '(un)trashed_post' hook twice (for parent and variation)
wp_trash_post($variation1_id);

wp_untrash_post($variation1_id);

$variation1 = wc_get_product($variation1_id);

// Fires actions for hooks 'trashed_post', 'woocommerce_trash_product_variation'
// (twice, for parent and variation, for each)
$variation1->delete(false);

wp_untrash_post($variation1_id);

// Fires actions for 'woocommerce_before_delete_product_variation' and 'delete_post' hooks
// (twice, for parent and variation, for each)
$variation1->delete(true);

[$parent, $variation1, $variation2] = create_test_variable_product();

// This fires the action multiple times:
//	For 'woocommerce_before_delete_product' hook (parent)
//	For 'woocommerce_before_delete_product_variation' hook (for each variation)
//	For 'woocommerce_before_delete_product_variation' hook (parent, with variation_id's)
//	For 'delete_post' hook (for each variation)
//	For 'delete_post' (for the parent, with variation_id's)
//	For 'delete_post' (for the parent parent)
$parent->delete(true);

[$parent, $variation1, $variation2] = create_test_variable_product();

// Same as above minus the action for the 'woocommerce_before_delete_product' hook

wp_delete_post($parent->get_id(), true);
```

#### 6. SQL level hooks

```php
$product = create_test_simple_product();
$product->set_manage_stock( true );
$product->set_stock_quantity( 100 );
$product->save();

// Fires action for the 'woocommerce_update_product' hook
wc_update_product_stock( $product->get_id(), 50, 'set' );

[$parent, $variation1, $variation2] = create_test_variable_product();

// Fires action for the ' woocommerce_updated_product_price' hook
WC_Product_Variable::sync($parent->get_id, false);

$data_store = WC_Data_Store::load( 'product' );

// Fires action for the ' woocommerce_updated_product_sales' hook
$data_store->update_product_sales( $parent->get_id(), 5, 'increase' );
```

#### 7. Direct invalidations

```php
$parent->set_name( 'Renamed Sync Test Product' );

// Fires actions for the 'save_post_product' and 'woocommerce_update_product' hooks,
// and for the 'WC_Product_Variable_Data_Store_CPT::sync_variation_names' method
// (for parent and variations)
$parent->save();

$parent->set_manage_stock(true);
$parent->set_stock_status( 'outofstock' );

// Fires actions for the 'woocommerce_update_product' hook,
// and for the 'WC_Product_Variable_Data_Store_CPT::sync_managed_variation_stock_status ' method
// (for parent and variations)
$parent->save();
```

#### 8. Changes in attributes

For this block we need an extra snippet in the server:

```php
function create_test_attribute_and_product() {
	delete_transient( 'wc_attribute_taxonomies' );

	$attr_args = array(
		'name'         => 'Test Attr ' . time(),
		'slug'         => 'test-attr-' . time(),
		'type'         => 'select',
		'order_by'     => 'menu_order',
		'has_archives' => false,
	);
	$attr_id = wc_create_attribute( $attr_args );
	if ( is_wp_error( $attr_id ) ) {
		echo "✗ FAIL: Could not create test attribute: " . $attr_id->get_error_message() . "\n";
		return;
	}

	$taxonomy = wc_attribute_taxonomy_name( $attr_args['slug'] );
	register_taxonomy( $taxonomy, array( 'product' ), array( 'hierarchical' => false ) );

	$term = wp_insert_term( 'Test Value', $taxonomy );
	if ( is_wp_error( $term ) ) {
		echo "✗ FAIL: Could not create test term: " . $term->get_error_message() . "\n";
		wc_delete_attribute( $attr_id );
		return;
	}
	$term_id = $term['term_id'];

	$variable_product = new WC_Product_Variable();
	$variable_product->set_name( 'Attribute Test Product' );

	$attribute = new WC_Product_Attribute();
	$attribute->set_id( $attr_id );
	$attribute->set_name( $taxonomy );
	$attribute->set_options( array( $term_id ) );
	$attribute->set_visible( true );
	$attribute->set_variation( true );
	$variable_product->set_attributes( array( $attribute ) );
	$variable_product->save();

	$term_obj  = get_term( $term_id, $taxonomy );
	$variation = new WC_Product_Variation();
	$variation->set_parent_id( $variable_product->get_id() );
	$variation->set_attributes( array( $taxonomy => $term_obj->slug ) );
	$variation->set_regular_price( '10.00' );
	$variation->save();

	return [
		$attr_id,
		$term_id,
		$taxonomy,
		$variable_product,
		$variation
	];
}
```

And here's the tests block:

```php
[$attr_id, $term_id, $taxonomy, $variable_product_id, $variation] = create_test_attribute_and_product();

// Fires action for 'woocommerce_attribute_updated' hook
wc_update_attribute( $attr_id, array( 'name' => 'Renamed Test Attr' ) );

// Fires action for 'edited_term' hook
wp_update_term( $term_id, $taxonomy, array( 'name' => 'Renamed Value' ) );

// Fires action for 'woocommerce_attribute_updated' hook
wc_update_attribute( $attr_id, array( 'name' => 'Renamed Test Attr' ) );

// Fires action for 'woocommerce_attribute_deleted' hook
wc_delete_attribute( $attr_id );

// Fires action for 'woocommerce_updated_product_attribute_summary' hook twice
// (parent and variation)
do_action( 'woocommerce_updated_product_attribute_summary', $variation->get_id() ); 
```

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

joshuatf committed:
Add initial cache invalidation interface and product invalidator

joshuatf committed:
Update product invalidation hooks to fire on post transition

joshuatf committed:
Fix product variation creation not triggering invalidation

joshuatf committed:
Keep operations consistent

joshuatf committed:
Clean up unnecessary context

joshuatf committed:
Update invalidator to use WP hooks instead of WC hooks

joshuatf committed:
Add initial tests

joshuatf committed:
Use WC hooks for variations instead of WP hooks

joshuatf committed:
Add changelog entry

ObliviousHarmony commented on plugins/woocommerce/src/Caches/Invalidators/CacheInvalidatorInterface.php:
```diff
@@ -0,0 +1,32 @@
+	 */
+	const OPERATION_CREATE = 'create';
+	const OPERATION_UPDATE = 'update';
+	const OPERATION_DELETE = 'delete';
```
Out of curiosity, what do you see as the benefit of having an `$operation` string associated with the invalidation event?

joshuatf replied:
It may or may not influence the subscriber's decision on how to handle changes to cache.  For example, it's feasible that a caching mechanism manually maintains a cached state of entities.  Knowing what the action is allows the subscriber to update, add, or delete something from their cache respectively instead of regenerating the entire cached array of entities.

That's not always how it works in practice, but the context is a nice addition that may come in handy.

ObliviousHarmony commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -0,0 +1,177 @@
+	 *
+	 * @return void
+	 */
+	private function register_hooks(): void {
```
While this is a great start, we should probably be more aggressive about detecting possible invalidating events. The `product.updated` and `product.created`  web hooks seem to have the `woocommerce_process_product_meta` present as well.

Should we considering trying to catch postmeta updates for product-related changes?

If we're going to ask people to rely on the behavior it should be comprehensive!

joshuatf replied:
Correct me if I'm wrong, but doesn't the `woocommerce_process_product_meta` hook just run on post save?  https://github.com/woocommerce/woocommerce/blob/de86daf1068901d46255d5776c3c27975cf1d1cf/plugins/woocommerce/includes/admin/class-wc-admin-meta-boxes.php/#L49

I don't think it's going to catch anything else that hasn't been added here, but please let me know if you see a case that slips by the current post save check.

ObliviousHarmony replied:
Yep, you're totally right!

I have taken a look at WooCommerce Subscriptions and WooCommerce Bookings to see how they modify post data and whether or not the current hooks are comprehensive enough:

- `update_post_meta()` can be called directly to change product metadata.
- [`wp_insert_post()` actually runs _before_ the product model is fully initialized](https://github.com/woocommerce/woocommerce/blob/e50d54857f051aaa0a71567f47a6890e9a78b787/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L249-L255). Correct me if I'm am wrong, however, this means that the invalidation hook will run on a half-initialized model, right? We should probably use `woocommerce_new_product` and `woocommerce_update_product` since those run as part of the product save lifecycle?
- `wp_update_post()` [is only called if post data is updated](https://github.com/woocommerce/woocommerce/blob/e50d54857f051aaa0a71567f47a6890e9a78b787/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L322).
- Our `save_post` hook should be a lower priority so that we let extensions and others run before we trigger the invalidation. I took a look at some major extensions and they hook `save_post` to update metadata.
- Even in WooCommerce, we have cases where we use direct DB queries to update post meta without saving. [You can update stock with a query](https://github.com/woocommerce/woocommerce/blob/e50d54857f051aaa0a71567f47a6890e9a78b787/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L1682), as an immediate example.

ObliviousHarmony commented on plugins/woocommerce/tests/php/src/Caches/Invalidators/ProductCacheInvalidatorTest.php:
```diff
@@ -0,0 +1,184 @@
+	/**
+	 * @testdox Implements CacheInvalidatorInterface.
+	 */
+	public function test_implements_interface() {
```
This test feels super unnecessary.

joshuatf replied:
Agreed.  This test can be removed.

ObliviousHarmony commented on plugins/woocommerce/tests/php/src/Caches/Invalidators/ProductCacheInvalidatorTest.php:
```diff
@@ -0,0 +1,184 @@
+		$parent_product = \WC_Helper_Product::create_variation_product();
+		$variations     = $parent_product->get_children();
+
+		// Should have create event for parent + create events for each variation + update events for parent for each variation.
```
Should our test make sure this is all true?

joshuatf replied:
Good call.  Updated in https://github.com/woocommerce/woocommerce/pull/61613/commits/5c15078b924f712c755a54a022a8a1278e7332f1.

ObliviousHarmony commented on plugins/woocommerce/tests/php/src/Caches/Invalidators/ProductCacheInvalidatorTest.php:
```diff
@@ -0,0 +1,184 @@
+	/**
+	 * @testdox Invalidating a product fires woocommerce_product_cache_invalidated action with correct parameters.
+	 */
+	public function test_invalidate_method() {
```
This test seems superfluous given that all of the other tests already hit this code path. What do you think?

joshuatf replied:
Mostly agreed.  There's enough coverage that we don't really need this test at the moment.  However, it is a public method that can be directly called.

I like to leave test coverage for public API in place in case there are future changes to the underlying mechanisms that end up removing the indirect usage of the method and leaving us with less coverage.

ObliviousHarmony replied:
It doesn't hurt :smile:

ObliviousHarmony reviewed (commented):
Thanks @joshuatf,

I _really_ like the architecture you've developed here. It's a convenient way to collapse all of the different hooks and instance methods down into a single filter to bust the cache of a product. I've ran through your testing instructions and it seems to work exactly as expected.

One concern that I have about this approach is how frequently `woocommerce_product_cache_invalidated` can be called. We have this action so that consumers can detect when the cache is invalidated, but technically, a cache is only invalidated if it was updated in the first place. It might just be the name that I'm leaning against here, but we should be careful about how we steer people to use this. For example, if someone takes this as an invitation to update the cache, complicated multi-step updates across extensions could lead to performance issues.

Do you think I'm overthinking that?

joshuatf committed:
Update tests based on feedback

joshuatf committed:
Fix up missing wp_post arg in delete hook

joshuatf commented:
Thanks for the feedback @ObliviousHarmony!  Made some of the suggestions.

>One concern that I have about this approach is how frequently `woocommerce_product_cache_invalidated` can be called. We have this action so that consumers can detect when the cache is invalidated, but technically, a cache is only invalidated if it was updated in the first place.

I think ultimately it's up to the subscriber to determine whether or not they need to invalidate their own cache or do nothing.  We should notify subscribers as much as possible when updates are made.  But...

>It might just be the name that I'm leaning against here, but we should be careful about how we steer people to use this. For example, if someone takes this as an invitation to update the cache, complicated multi-step updates across extensions could lead to performance issues.

I do agree that the name could be better.  We aren't really the cache handler in this case and don't necessarily know that it needs to be invalidated.

I would lean into a name with "publisher" like `EntityEventPublisher` though I'm wary of those names given WP has a tendency to overuse those terms as is.

Maybe something like `EntityEventNotifier` or `EntityEventDispatcher`.  Thoughts?

ObliviousHarmony reviewed (dismissed):
> Maybe something like EntityEventNotifier or EntityEventDispatcher. Thoughts?

I like `EntityEventDispatcher`.

---

To be honest, this PR and my findings about missed cache invalidation events really highlights how hard invalidation can be. Providing a central location for listening to invalidation events and supporting invalidation is a great solution to a whole slew of problems with it.

With that said, if we do this, it _has_ to be comprehensive. Once people begin to use it they will stop worrying about invalidation. When that happens, any time changes are made to entities that aren't caught, we are going to potentially introduce bugs. For example, we might start serving stale REST responses from cache for product data that has been changed.

I think we need to look very broadly at the ecosystem and validate any assumptions about what will and won't require invalidation events. That'll tell us if it's even safe enough to add a layer like this in the first place.

Konamiman commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -357,6 +358,7 @@ private function init_hooks() {
 		$container->get( AddressProviderController::class );
 		$container->get( AbilitiesRegistry::class );
 		$container->get( MCPAdapterProvider::class );
+		$container->get( ProductCacheInvalidator::class );
```
`init` is automatically executed when the DI container creates the instance of the class, so this is not an actual issue.

Konamiman committed:
Merge branch 'trunk' into add/cache-invalidator

Konamiman committed:
Significant implementation changes to ProductCacheInvalidator:

- CacheInvalidatorInterface removed
- More product modification related hooks listened to,
  including attribute-related changes
- Added manual invalidation for direct SQL data store operations

Konamiman committed:
Fix linting issues

Konamiman committed:
Merge branch 'trunk' into add/cache-invalidator

Konamiman commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+<?php
+
+declare( strict_types=1 );
```
The `$invalidate` method has been modified so `$product_id` is always an integer.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+		 *                           - 'old_slug' (string) - Previous attribute slug. Present for attribute renames.
+		 *                           - 'new_slug' (string) - New attribute slug. Present for attribute renames.
+		 */
+		do_action( 'woocommerce_product_cache_invalidated', $product_id, $operation, $context );
```
Good thing to note in documentation.

Konamiman replied:
This hook has been removed.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+	 *
+	 * @return void
+	 */
+	private function invalidate_products_with_attribute( string $taxonomy, array $context = array() ): void {
```
This method will be used only when the products data store in use is the standard CPT one provided by WooCommerce, so we're all set.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+	 *
+	 * @since 10.5.0
+	 */
+	public function invalidate( $product_id, string $operation, array $context = array() ): void {
```
With the change of the `invalidate` signature to force `$product_id` to be an `int` this is no longer an issue.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+	 *
+	 * @since 10.5.0
+	 */
+	final public function init(): void {
```
By convention all the `init` methods of classes resolved via DI container are `final`, and ve even have a linting rule enforcing that.

Konamiman commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+	 * @since 10.5.0
+	 */
+	public function invalidate( $product_id, string $operation, array $context = array() ): void {
+		wc_get_container()->get( VersionStringGenerator::class )->delete_version( "product_{$product_id}" );
```
The DI container is always available once WooCommerce has been initialized and we know that the `VersionStringGenerator` class will always be available. We don't have error handling around each and every usage of the DI container.

Konamiman committed:
Force product_id in ProductCacheInvalidator::invalidate to always be an int

prettyboymp commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+		}
+
+		if ( 'product_variation' === $post->post_type ) {
+			$variation = wc_get_product( $post_id );
```
Can we avoid calling `wc_get_product()`?  Wouldn't the `$post` have the parent?

Konamiman replied:
Good catch. I have modified the code to use either the available parent id or a call to `wp_get_post_parent_id` instead of retrieving the full product object.

prettyboymp commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+		$this->invalidate( $variation_id, $operation, $context );
+
+		if ( is_null( $variation ) ) {
+			$variation = wc_get_product( $variation_id );
```
Same here, it would get great to avoid the overhead of calling `wc_get_product()`.

prettyboymp commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,725 @@
+
+		$product_ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT DISTINCT post_id FROM {$wpdb->postmeta}
```
I have concerns about the performance of these queries on large tables with sites that rely heavily on attributes.  Are the post_ids already cached via the term handling?

Konamiman replied:
I have modified the code to use the term relationships table (indexed) when invalidating products associated to a given term, and to use the WP object cache to store short lived lists of products associated to terms or taxonomies (the TTL of these cached entries is filterable).

prettyboymp reviewed (commented):
Overall, I like the idea of simplifying hooks that need to be dealt with to capture all changes to a product.  My concern, though, is performance... because we're capturing ALL changes to a product.  Some of these will fire multiple times when a product is saved, which itself, can happen dozens of times during a process like checkout.

I did some local testing with k6, going through checkout with different product types/subscriptions, and didn't find any slow down from having this active, at least not outside of standard deviation.  So, I think the performance impact is minor at this point.  But, we will need to be extremely careful moving forward.  We should do everything we can to avoid one of these handlers introducing a query or any other expensive check.

Konamiman committed:
Optimize the flow of parent+variation invalidations

Konamiman committed:
Optimize the cache invalidation for term/attribute changes:

- Query the term relationships table (indexed) for term related invalidations
- Cache the list of products related to a term/attribute (filterable TTL)

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,798 @@
+	 *
+	 * @return void
+	 */
+	private function invalidate_products_with_attribute( string $taxonomy, array $context = array() ): void {
```
This method will only be executed when the CPT data store is in place (see `register_hooks` method).

Konamiman committed:
Rename ProductCacheInvalidator to ProductVersionStringInvalidator and make it internal

Konamiman committed:
Remove the woocommerce_product_cache_invalidated action

Also remove the now unused arguments $operation and $context in the invalidate method.

Konamiman committed:
Merge branch 'trunk' into add/cache-invalidator

Konamiman committed:
Remove unused action handler arguments

Konamiman committed:
Remove the MANUAL_TESTING file

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/MANUAL_TESTING.md:
```diff
@@ -0,0 +1,1128 @@
+        'function' => $context['function'] ?? null,
+        'context_keys' => array_keys( $context )
+    ];
+}, 10, 3 );
```
This file has been deleted, it was committed by mistake.

Konamiman commented on plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php:
```diff
@@ -749,6 +750,13 @@ public function sync_variation_names( &$product, $previous_name = '', $new_name
+
+			$invalidator = wc_get_container()->get( ProductVersionStringInvalidator::class );
+			$children    = $product->get_children();
+			foreach ( $children as $child_id ) {
```
The class is not expected to be removed, and if it is, the client code will be updated accordingly.

Konamiman commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -0,0 +1,588 @@
+	 *
+	 * @param FeaturesController $features_controller The features controller.
+	 *
+	 * @return void
```
Being inside the `Automattic\WooCommerce\Internal` namespace already makes the class internal, no need for an extra annotation.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,798 @@
+	 *
+	 * @since 10.5.0
+	 */
+	public function invalidate( int $product_id, string $operation, array $context = array() ): void {
```
This class is now in the `Internal` namespace, so not intended for usage outside of WooCommerce core.

Konamiman commented on plugins/woocommerce/src/Caches/Invalidators/ProductCacheInvalidator.php:
```diff
@@ -0,0 +1,798 @@
+		 *                           - 'old_slug' (string) - Previous attribute slug. Present for attribute renames.
+		 *                           - 'new_slug' (string) - New attribute slug. Present for attribute renames.
+		 */
+		do_action( 'woocommerce_product_cache_invalidated', $product_id, $operation, $context );
```
This action has been removed.

albarin reviewed (approved):
Changes are looking good.

✅ Confirmed all 8 testing cases work as described.

Konamiman committed:
Remove type hints from hook handlers for robustness

Konamiman commented on plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php:
```diff
@@ -749,6 +750,13 @@ public function sync_variation_names( &$product, $previous_name = '', $new_name
 				)
 			);
+
+			$invalidator = wc_get_container()->get( ProductVersionStringInvalidator::class );
```
> ensure the ProductVersionStringInvalidator::invalidate() method has long-term stability guarantees.

It has. If we change or move the method we'll update the client code accordingly (it's all in the same codebase).

Konamiman commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -357,6 +358,7 @@ private function init_hooks() {
 		$container->get( AddressProviderController::class );
 		$container->get( AbilitiesRegistry::class );
 		$container->get( MCPAdapterProvider::class );
+		$container->get( ProductVersionStringInvalidator::class );
```
> Plan for long-term compatibility

That's assumed.

Konamiman committed:
Rename woocommerce_cache_invalidator_taxonomy_lookup_ttl to woocommerce_version_string_invalidator_taxonomy_lookup_ttl

Also change the swecond parameter from class name to 'product' (entity type)

Konamiman committed:
Merge branch 'trunk' into add/cache-invalidator

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Caches/ProductVersionStringInvalidator.php:
```diff
@@ -0,0 +1,603 @@
+	 * @internal
+	 */
+	final public function init( FeaturesController $features_controller ): void {
+		if ( ! $features_controller->feature_is_enabled( 'rest_api_caching' ) ) {
```
@joshuatf this line leads to `FeaturesController::init_feature_definitions()` ([ref](https://github.com/woocommerce/woocommerce/blob/508cf96e8f7d0685b3f7ddca1615e3c2016bf898/plugins/woocommerce/src/Internal/Features/FeaturesController.php#L286-L288)) being called and for some translations to be loaded, effectively while loading plugins.

That leads to the dreadful deprecation notice on every PHP load:

> Function `_load_textdomain_just_in_time` was called <strong>incorrectly</strong>. Translation loading for the <code>woocommerce</code> domain was triggered too early. This is usually an indicator for some code in the plugin or theme running too early. Translations should be loaded at the <code>init</code> action or later. Please see <a href="https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 6.7.0.)

Can you open a follow-up PR to address this by checking for the feature later, or I should open an issue for that?

Konamiman replied:
Hi, I was actually in charge of this pr since Joshua left Automattic. Yes, please open an issue for this.

RadoslavGeorgiev replied:
This is my second strike trying to ping him! 🤦 

Issue WOOPLUG-6053 and PR https://github.com/woocommerce/woocommerce/pull/62541 opened.
